src/test/java/org/opensearch/plugin/insights/core/service/QueryGroupingServiceTests.java

```diff
@@ -0,0 +1,1041 @@
+            }
+        }
+
+        assertEquals(2, queryGroupingService.numberOfTopGroups());
+        assertEquals(900L, topQueriesStore.poll().getMeasurement(MetricType.LATENCY));
+        assertEquals(1000L, topQueriesStore.poll().getMeasurement(MetricType.LATENCY));
+    }
+
+    // 6. Existing query group update to MIN decrease average - overflows to MAX
+    public void testExistingGroupUpdateToMinDecreaseAverageOverflowsToMax() {
+        queryGroupingService = new QueryGroupingService(
+            MetricType.LATENCY,
+            GroupingType.SIMILARITY,
+            DimensionType.AVERAGE,
+            topQueriesStore,
+            2
+        );
+        int numOfRecords = 1;
+
+        List<SearchQueryRecord> records1 = QueryInsightsTestUtils.generateQueryInsightsRecordsWithMeasurement(
+            numOfRecords,
+            MetricType.LATENCY,
+            1000
+        );
+        // Set all records to have the same hashcode for aggregation
+        QueryInsightsTestUtils.populateHashcode(records1, 1);
+
+        List<SearchQueryRecord> records2 = QueryInsightsTestUtils.generateQueryInsightsRecordsWithMeasurement(
+            numOfRecords,
+            MetricType.LATENCY,
+            1100
+        );
+        // Set all records to have the same hashcode for aggregation
+        QueryInsightsTestUtils.populateHashcode(records2, 2);
+
+        List<SearchQueryRecord> records3 = QueryInsightsTestUtils.generateQueryInsightsRecordsWithMeasurement(
+            numOfRecords,
+            MetricType.LATENCY,
+            1199
+        );
+        // Set all records to have the same hashcode for aggregation
+        QueryInsightsTestUtils.populateHashcode(records3, 3);
+
+        // Average will decrease when this record added and record should be moved out of Top N
+        List<SearchQueryRecord> records4 = QueryInsightsTestUtils.generateQueryInsightsRecordsWithMeasurement(
+            numOfRecords,
+            MetricType.LATENCY,
+            1
+        );
+        // Set all records to have the same hashcode for aggregation
+        QueryInsightsTestUtils.populateHashcode(records4, 3);
+
+        SearchQueryRecord aggregatedRecord = null;
+        List<List<SearchQueryRecord>> allRecords = Arrays.asList(records1, records2, records3, records4);
+
+        for (List<SearchQueryRecord> recordList : allRecords) {
+            for (SearchQueryRecord record : recordList) {
+                aggregatedRecord = queryGroupingService.addQueryToGroup(record);
+            }
+        }
+
+        assertEquals(2, queryGroupingService.numberOfTopGroups());
+        assertEquals(1000L, topQueriesStore.poll().getMeasurement(MetricType.LATENCY));
+        assertEquals(1100L, topQueriesStore.poll().getMeasurement(MetricType.LATENCY));
+    }
+
+    // 7. Existing query group update to MAX increases average - stay in MAX
+    public void testExistingGroupUpdateToMaxIncreaseAverageStayInMax() {
+        queryGroupingService = new QueryGroupingService(
```

Comment:
A lot of the code in those cases are the same set up. Please simplify with a shared function.